src/fugw/solvers/utils.py

```diff
@@ -996,21 +1216,29 @@ def compute_quad_kl_sparse(mu, nu, alpha, beta):
 
 
 def compute_quad_l2(a, b, mu, nu):
-    """
-    Compute || a otimes b - mu otimes nu ||^2
-    """
+    """Compute || a otimes b - mu otimes nu ||^2."""
 
-    norm_a2 = (a**2).sum()
-    ratio = (a * mu).sum() / norm_a2
-    norm = norm_a2 * torch.sum((b - ratio * nu) ** 2) + (nu**2).sum() * (
-        torch.sum((mu**2)) - ratio
+    norm = (
+        (a**2).sum() * (b**2).sum()
+        - 2 * (a * mu).sum() * (b * nu).sum()
+        + (mu**2).sum() * (nu**2).sum()
     )
 
     return norm
 
 
-def compute_l2(p, q):
-    return torch.sum((p - q) ** 2)
+def compute_quad_l2_sparse(a, b, mu, nu):
+    """Compute || a otimes b - mu otimes nu ||^2."""
+
```

Comment:
Same remark + no check of sparsity pattern ?

Reply:
We could add sparsity checks indeed, but these functions are not meant to be public so I guess we are somewhat safe.